vxfw/vxlayout/flex2.go

```diff
@@ -0,0 +1,422 @@
+	"os"
+
+	"git.sr.ht/~rockorager/vaxis"
+	"git.sr.ht/~rockorager/vaxis/vxfw"
+)
+
+// Flexible describes a [vxfw.Widget] that can take a flexible amount of space in a [Row] or
+// [Column].
+type Flexible interface {
+	vxfw.Widget
+	FlexFactor() uint16
+	FlexLoose() bool
+}
+
+// Determines how children in a [Row] or [Column] are laid out on the cross axis.
+// The default is [CrossAxisCenter], which centers children in the available cross axis space.
+// Use [CrossAxisStart] to align children to the top of a [Row] or left of a [Column], and vice
+// versa for [CrossAxisEnd].
+// [CrossAxisStretch] will force all children to fill the maximum space on the cross axis.
+type CrossAxisAlignment int
+
+const (
+	CrossAxisCenter CrossAxisAlignment = iota
+	CrossAxisStart
+	CrossAxisEnd
+	CrossAxisStretch
+)
+
+// Determines how children in a [Row] or [Column] are laid out on the main axis.
+// The default is [MainAxisStart], which places children at the left or top of a [Row] or [Column]
+// respectively.
+// Use [MainAxisCenter] to center children, or one of the Space variants to distribute the space
+// elsewhere.
+// Note that these options do nothing if one of the children has a flex factor > 0, as those
+// children will take the available space.
+type MainAxisAlignment int
+
+const (
+	MainAxisStart MainAxisAlignment = iota
+	MainAxisEnd
+	MainAxisCenter
+	MainAxisSpaceBetween
+	MainAxisSpaceAround
+	MainAxisSpaceEvenly
+)
+
+type LayoutOptions struct {
+	MainAxis  MainAxisAlignment
+	CrossAxis CrossAxisAlignment
+
+	// Gap controls how much space is placed between each child *before* the children are sized.
+	Gap uint16
+}
+
+// Row returns a [vxfw.Widget] that lays out children horizontally.
+func Row(children []vxfw.Widget, options LayoutOptions) vxfw.Widget {
+	return &flex{children: children, options: options, direction: FlexHorizontal}
+}
+
+// Column returns a [vxfw.Widget] that lays out children vertically.
+func Column(children []vxfw.Widget, options LayoutOptions) vxfw.Widget {
+	return &flex{children: children, options: options, direction: FlexVertical}
+}
+
+type box struct {
+	vxfw.Widget
+	flex  uint16
+	loose bool
+}
+
+var (
+	_ vxfw.Widget = box{}
+	_ Flexible    = box{}
+)
+
+func (b box) Draw(ctx vxfw.DrawContext) (vxfw.Surface, error) {
+	return b.Widget.Draw(ctx)
+}
+
+func (b box) FlexLoose() bool    { return b.loose }
+func (b box) FlexFactor() uint16 { return b.flex }
+
+// Expanded returns a [vxfw.Widget] that will expand in a flexible layout based on flex.
+func Expanded(widget vxfw.Widget, flex uint16) vxfw.Widget {
+	return box{Widget: widget, flex: flex}
+}
+
+// Flex returns a [vxfw.Widget] that will loosely flex in a flexible layout based on flex.
+func Flex(widget vxfw.Widget, flex uint16) vxfw.Widget {
+	return box{Widget: widget, flex: flex, loose: true}
+}
+
+type widgetFunc func(vxfw.DrawContext) (vxfw.Surface, error)
+
+func (w widgetFunc) Draw(ctx vxfw.DrawContext) (vxfw.Surface, error) {
+	return w(ctx)
+}
```

Comment:
I'd like to move this up to package `vxfw` as `vxfw.WidgetFunc`. As you can see in some of the below widgets it's a really nice API to create stateless widgets.

Reply:
I second this with the addition of a stateless event handler, making a struct having multiple widgets, and each widget having their own event handler.

Reply:
I'll play around with it in a separate PR and folks can see how it feels.

Reply:
https://github.com/rockorager/vaxis/pull/26